Reader: readability, autosizing and performance.

kurzee opened the pull request:
Welcome to the Reader readability, autosizing and scrolling performance PR. The bulk of this PR is geared towards improving our Reader experience on iPad and preparing the way for Reader split-view support. However, many many changes had to be made to support this and keep the experience optimal on both iPhone and iPad:

- Removed a ton of technical debt/code.
- Removed all of the previous manual cell height and layout calculation code.
- Restructured all of our Reader cells for autosizing.
- Additionally restructured our cells with `UIStackViews` for ease of layout and future adaptability.
- Following Apple's ordained readable margins on all of our cell's content for both iPhone and iPad.
- Took a step toward the upcoming Reader card refresh design by extending our borders to the edges of the screen rather than the boxes we've had before. This is also a far less complicated layout == better scrolling performance.
- Paid very close attention to scrolling performance on all devices, including iPhone 4S, attempting to improve performance where we can.
- Improved refresh, appearance and back-navigation performance.

**Note:** The `ReaderDetailViewController` was only minimally touched for this PR and will await the upcoming detail view refresh to receive a needed layout overhaul.

**Note:** The `ReaderCommentsViewController` was barely touched on this PR and will need its own revamp to properly support readability, autosizing and iPad. For now, we present Reader comments as a modal on iPad.

**Bonus:**
Fixes: #5994 
Fixes: #5981 

## iPhone and iPad
1. Ensure Reader cards render as expected on all Reader menu streams.
2. Ensure Reader stream scrolling performance is as expected and at least as good as before on older devices, such as iPhone 4S.
3. Ensure Reader cards are following readable margins on iPhone, iPhone Plus landscape, and iPad.
4. Test that all buttons on Reader cards work as expected.

## iPad
1. Check that comments open in a modal view controller for Reader streams and from Reader detail.
2. Check that #5082 doesn't reproduce.
3. Test transitioning from iPad split-multitasking transitions nicely and correctly for readable margins.

## iPhone
1. Check that Reader comments open as a push onto the navigation stack.
2. Check that #5082 doesn't reproduce on iPhone Plus landscape orientation.

## Review

- @frosty are you willing to take a long stroll through some Reader code? I promise most of the `+-` are from nibs. 😅 We can add additional code reviewers if you'd like.
- @aerych can you take a look at the changes made in `WPTableViewHandler` as those are particular to the many discussions we've had.

Also, @aerych it would be amazing if you could spend some time testing throughout the Reader experience on both iPhone and iPad to ensure I haven't missed any gotchas or other historical Reader issues that you would be familiar with.

## Design Review
@jancavan and/or @mattmiklic can you take a stroll through this build or screenshots? We've made some scrolling performance-dependent design decisions in sync with the upcoming cards refresh. I opted not to go too far into the new refresh design or into what is requested on #5769 as those are better suited for a separate PR. Let me know if anything stands out for this round that we should address now.

## Screenshots

![simulator screen shot oct 14 2016 5 25 12 pm](https://cloud.githubusercontent.com/assets/1873422/19404395/4c43b6d4-9233-11e6-8133-037461817839.png)

![simulator screen shot oct 14 2016 5 26 36 pm](https://cloud.githubusercontent.com/assets/1873422/19404410/6d1ad298-9233-11e6-9f45-d43c2eb9936c.png)

aerych commented:
> Also, @aerych it would be amazing if you could spend some time testing throughout the Reader experience on both iPhone and iPad to ensure I haven't missed any gotchas or other historical Reader issues that you would be familiar with.

![im-your-huckleberry](https://cloud.githubusercontent.com/assets/1435271/19404981/6220f5f2-9238-11e6-9e78-0ad2f7b8f758.gif)

frosty commented:
Amazing work, @kurzee! 😅 I've done some initial testing in the simulators, and I've found a few UI issues – some of this may be existing issues, but I still wonder if they're things we could address here if we're already rebuilding the reader cards:

* Images on the cards are no longer full width – they have margins on either side. Is this intentional? I don't have a strong opinion either way, really – _perhaps_ the full width images make better use of the space?
* There's no longer an ellipsis at the end of the content preview text (it just cuts off after the last word, rather than ending '...')
* In Discover, the attribution label ('visit xyz', 'originally posted on...' no longer begins with a capital letter.
* Non WordPress.com posts (or, those without tags, comments, and likes), sometimes leave extra padding at the bottom of their cells (as though for the likes / comments buttons), and sometimes don't. See these two cells, how their padding is different: 

![simulator screen shot 17 oct 2016 16 15 10](https://cloud.githubusercontent.com/assets/4780/19443615/28ffbef8-9485-11e6-95ce-151cd2e769fd.png)

- (Existing issue) The comments button doesn't show up on some posts that I know have comments. For example, in the Followed Sites list ma.tt posts don't show the comments button, but they do in the Site Details view:

![reader-comments-button](https://cloud.githubusercontent.com/assets/4780/19443725/92007802-9485-11e6-8189-ae4f42a702d1.png)

* The Fetching Posts animation doesn't appear in landscape on iPhone ('normal' and + size) – just the text.
* I saw one or two instances of placeholder blavatars not displaying, or being the wrong width. I'm not sure how to reproduce this (worth checking MacRumors):

![reader-blavatar-size](https://cloud.githubusercontent.com/assets/4780/19443920/518a272c-9486-11e6-8aa3-0fe9d7a66283.png)

* 'Be the first to leave a comment' on the iPad comment modal doesn't stay vertically centered
* Post headings are still limited to 4 lines – should be 0 (#5994)
* The image on the first card changes when cell is recycled (and looks better after being recycled). Here's an example, when I first navigate into a tag:

![simulator screen shot 17 oct 2016 11 53 39](https://cloud.githubusercontent.com/assets/4780/19444041/c923a40c-9486-11e6-8d07-52df1d347a70.png)

And then after recycling the cell:

![simulator screen shot 17 oct 2016 11 53 51](https://cloud.githubusercontent.com/assets/4780/19444059/dd3b956c-9486-11e6-9b3e-4c6b5166aee9.png)

I saw the same issue occur when changing the app's size with multitasking – the image's content would be incorrect until it was recycled. Before:

![simulator screen shot 17 oct 2016 15 34 12](https://cloud.githubusercontent.com/assets/4780/19444106/00421eaa-9487-11e6-8544-394a5fc89b16.png)

After:

![simulator screen shot 17 oct 2016 15 34 25](https://cloud.githubusercontent.com/assets/4780/19444109/049e57ca-9487-11e6-98a3-767825078ac4.png)

* I saw one instance where the word count label was collapsed vertically, showing only a thin sliver of content. I'll grab you a screenshot of this one shortly. Haven't been able to reproduce it yet.
* I think it looks kind of weird when a post has a word count, comments, and likes, but no tags. Would it be possible to stack the word count and tags labels in a vertical stack view, so we can collapse it if only one of them exists?

![reader-wordcount-comments-likes](https://cloud.githubusercontent.com/assets/4780/19444281/b595133e-9487-11e6-9aab-8ffe26c78049.png)

I'm working my way through the code now, will reply back with any comments I have!

aerych commented:
> Images on the cards are no longer full width – they have margins on either side. Is this intentional? I don't have a strong opinion either way, really – perhaps the full width images make better use of the space?

If its convenient, let's keep the margins so the image is flush left and right with the copy.  This will better match some upcoming changes in the web reader we'll also be adopting in the app.

aerych commented:
> For now, we present Reader comments as a modal on iPad.

Could you comment on why the comments vc needed to be moved to a modal?  Just curious what issue it was addressing.

kurzee commented:
>Could you comment on why the comments vc needed to be moved to a modal? Just curious what issue it was addressing.

That is a step towards support for split-view layout, since the comments vc cells rely on the ole fixed width hack.

kurzee commented:
>>Could you comment on why the comments vc needed to be moved to a modal? Just curious what issue it was addressing.

>That is a step towards support for split-view layout, since the comments vc cells rely on the ole fixed width hack.

@aerych that may be out of context for this PR so I can revert if needed. Alternatively, for iPad split-view, we could also force comments to be full-screen and never within a compact pane. Maybe something best left for @frosty's upcoming work.

aerych commented:
I spent some time testing this branch on a few different devices: 
iPad Pro (12") / iOS 9.3.5
iPad 3 / iOS 9.3.3
iPhone 5 / iOS 9.0.2

The post list and cards update was working great for me and I don't have too much to add there apart from what @frosty already noted in his excellent testing.  Here's what I do have:

- Scrolling on the iPad Pro is super smooth and the iPhone 5 has only a slight occasional stutter that seems to happen when an featured image is loaded.  However, scrolling on the iPad 3 is pretty rough when the stream is full of featured images. I'm not terribly concerned about it as the iPad 3 an older platform, and probably a bit under powered for iOS 9 and iOS 10.  I'd guess some of the stutter is due to memory pressure, e.g. images that are in-memory cache are purged in favor of new ones, or something of that nature.  If that's the case the smaller images in the upcoming cell refresh will help.  Still, maybe we can investigate options to improve scrolling performance for the iPad 3 in another PR. 

- I had one instance on the iPad 3 where I'd been browsing a list and backgrounded the app.  When I later resumed none of my taps on the cards would show a detail. I had to return to the menu, then open the list before I could view posts.  Not sure what's up with that and I can reproduce it on demand.  A random state restoration glitch I suppose, but mentioning it here for posterity.

- Again on the iPad 3, if I tap to select a topic with existing content (e.g. the Art tag), scroll down while the list is syncing, tapping the status bar to return to the top of the list once new content is loaded often results in a large empty gap at the top of the list until the next time I scroll. 

- Occasionally the loading view and the no results view are a bit off center, slightly too far down the screen, but I think this is an existing issue.

- Search:  One the iPad, the dropdown for past searches sticks out beyond the width of the new card content.  It might be nice if this could be made flush left and right.

That's about it for now. :)  I'm waiting for a nice back log of content in some of my followed feeds so I can see how a gap marker cell behaves and will follow up then if anything looks amiss.

aerych commented:
> @aerych that may be out of context for this PR so I can revert if needed. Alternatively, for iPad split-view, we could also force comments to be full-screen and never within a compact pane. Maybe something best left for @frosty's upcoming work.

If possible I'd be keen to not switch comments to a modal in one release, only for us to then switch it back to non-modal in another.  We'd just be inviting user confusion.  

That said, I care not one jot how we happen to shuffle it around while everything is still flowing through develop.  ;)

aerych commented:
> (Existing issue) The comments button doesn't show up on some posts that I know have comments. For example, in the Followed Sites list ma.tt posts don't show the comments button, but they do in the Site Details view:

The web reader avoids showing the comments button for Jetpack blogs, so for consistency we'd want to keep it hidden.  In the case of http://ma.tt, the reason the comment button appears in one place but not the other is due to differences in the fields being returned by the REST API between the `/read/following` endpoint and the `/read/sites` endpoint.  

On the other hand if we have the data why not show it and let the app lead the web. ;) 
Opened https://github.com/wordpress-mobile/WordPress-iOS/issues/6087 to follow up on this.

frosty commented:
>>> @aerych that may be out of context for this PR so I can revert if needed. Alternatively, for iPad split-view, we could also force comments to be full-screen and never within a compact pane. Maybe something best left for @frosty's upcoming work.

I'd prefer not to force comments to be full screen at the moment, as I still haven't found a suitable way to transition from a split to full screen whilst pushing / popping a VC.

kurzee commented:
Thanks @frosty, @aerych!!

Here's the first round of changes:

- Removed the word count label in https://github.com/wordpress-mobile/WordPress-iOS/pull/6077/commits/534ff2a796d52a7bfae3076c4429e95ff0b4dbec

>There's no longer an ellipsis at the end of the content preview text (it just cuts off after the last word, rather than ending '...')

- Fixed in https://github.com/wordpress-mobile/WordPress-iOS/pull/6077/commits/b2eaa19a49bafb1ef2fe8217499873e3e10a6ac6

>In Discover, the attribution label ('visit xyz', 'originally posted on...' no longer begins with a capital letter.

- Fixed in https://github.com/wordpress-mobile/WordPress-iOS/pull/6077/commits/e9e26e150ed23d1d1489564af0dfb93d0080de93

> Non WordPress.com posts (or, those without tags, comments, and likes), sometimes leave extra padding at the bottom of their cells (as though for the likes / comments buttons), and sometimes don't. See these two cells, how their padding is different:

- Fixed in https://github.com/wordpress-mobile/WordPress-iOS/pull/6077/commits/097b0477741ec10a340072219afb4ab6960f539c

> I saw one or two instances of placeholder blavatars not displaying, or being the wrong width. I'm not sure how to reproduce this (worth checking MacRumors):

- Fixed in https://github.com/wordpress-mobile/WordPress-iOS/pull/6077/commits/abc5ded38e3694b7b61ea0eea11afb9113364c19

> Post headings are still limited to 4 lines – should be 0 (#5994)

- Fixed in https://github.com/wordpress-mobile/WordPress-iOS/pull/6077/commits/1a3badf42d8d4bbd8bc85bda0dc9ded93e36b531

> If possible I'd be keen to not switch comments to a modal in one release, only for us to then switch it back to non-modal in another. We'd just be inviting user confusion.

- I've reverted the modal presentation change, for now, we'll tackle the comments issue as @frosty works on Reader split-view, @aerych. I think we'll probably have to go back to the modal on iPad though, as it will take some time to rework all of the comments cells for readability.

frosty commented:
Thanks for the changes, @kurzee! All your fixes look good. Continuing my code review now.

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderSiteStreamHeader.xib:
```diff
@@ -1,191 +1,121 @@
-<?xml version="1.0" encoding="UTF-8" standalone="no"?>
-<document type="com.apple.InterfaceBuilder3.CocoaTouch.XIB" version="3.0" toolsVersion="10117" systemVersion="15F34" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES">
+<?xml version="1.0" encoding="UTF-8"?>
```
I think the `didTapFollowButton` IBAction in here needs reconnecting – the method is called just fine, but IB shows it as a warning, and the .swift file shows the action as not connected.

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:
```diff
@@ -1592,60 +1531,34 @@ extension ReaderStreamViewController : WPTableViewHandlerDelegate {
     }
 
 
     public func tableView(aTableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
```
I'm not sure you need to implement this method any more, as you set `tableView.rowHeight` to `UITableViewAutomaticDimension` earlier.

kurzee replied:
Looks funky, I know, but we need the actual method since we're using `WPTableViewHandler` that asks for it via the method. However, I'm removing setting both `.estimatedRowHeight` and `.rowHeight` as it's unnecessary.

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:
```diff
@@ -236,17 +231,13 @@ import WordPressComAnalytics
         super.viewWillAppear(animated)
 
+        // Trigger layouts, if needed, to correct for any inherited layout changes, such as margins.
         refreshTableHeaderIfNeeded()
```
Do we not need to reload the tableview on `viewWillAppear` any more?

frosty commented on WordPress/Classes/ViewRelated/Reader/ReaderTagStreamHeader.xib:
```diff
@@ -1,110 +1,66 @@
-<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
The follow button IBAction is showing up as not connected properly – might just need disconnecting and reconnecting.

frosty commented on WordPress/Classes/ViewRelated/Reader/WPStyleGuide+Reader.swift:
```diff
@@ -142,31 +132,17 @@ extension WPStyleGuide
         let fontSize = Cards.buttonFontSize
```
Should we remove this method now we've ditched the word count?

kurzee commented:
>The image on the first card changes when cell is recycled (and looks better after being recycled). Here's an example, when I first navigate into a tag:
>I saw the same issue occur when changing the app's size with multitasking – the image's content would be incorrect until it was recycled. Before:

This is now fixed in https://github.com/wordpress-mobile/WordPress-iOS/pull/6077/commits/7b220415b96d945370e5072e37000e7fdcebe73a

To test: 
- Open the Reader on iPhone in landscape.
- Load a stream, such as Discover.
- Rotate the device to portrait, the image should resize/redownload as needed to match the new layout width for the imageview.
- Retest again, but starting with portrait and rotating to landscape on a different group of cells.

frosty commented:
A few more small Interface Builder issues:

* I'm seeing two issues when I first open `Reader.storyboard`. There's a conflicting constraint error in Reader Detail View Controller, and also if you select the container view in Reader it's far too big:

<img width="602" alt="screen shot 2016-10-18 at 10 23 15" src="https://cloud.githubusercontent.com/assets/4780/19471882/f8a62038-951c-11e6-91dc-64a147c411bf.png">

If you drag the bottom right grab handle and drag it up until the frame of the view snaps to the frame of the view controller, it seems to stay that way and doesn't introduce any warnings. Saving the storyboard like this, then navigating away from it and then re-opening it also seems to somehow fix the conflict error for me too.

* If you right click / ctrl click on the top level Reader Post Card Cell, there still seem to be a bunch of IBActions that don't match any existing methods any more (or, IB doesn't think they do). It seems like disconnecting and reconnecting fixes this.

kurzee commented:
>I'm seeing two issues when I first open Reader.storyboard. There's a conflicting constraint error in Reader Detail View Controller, and also if you select the container view in Reader it's far too big:

- Fixed the constraint issue in https://github.com/wordpress-mobile/WordPress-iOS/pull/6077/commits/90ad9e5aa4f741dec0db8f3efa1ee251d2a2e11f
 - I wasn't able to trigger the broken layout in the storyboard for the Reader container view, can you retry this one with maybe the latest commits? @frosty 

>If you right click / ctrl click on the top level Reader Post Card Cell, there still seem to be a bunch of IBActions that don't match any existing methods any more (or, IB doesn't think they do). It seems like disconnecting and reconnecting fixes this.

- Definitely can't figure out what's going on with these. Another day's rabbit hole. Or Xcode 8.1

## New Testing
For new commits deccfda9c1304d8c621d9ddd8b96e90277ee0f41 and b5af255448168f189e733c65aa7e286c8dccb2de, can you please test the following:
- Open Reader on iPad, starting in landscape orientation.
- Opening a stream of Reader cards.
- Take a screenshot of the first cells on the immediate load, at the top of the tableView.
- Scroll down and back to the top, to trigger a reuse of the cells.
- Take another screenshot and compare both screenshots.
- The layout should be exactly the same, exactly perfect, between the two screenshots of the same cells.

**State restoration:**

- Open Reader, open a stream.
- Background the app, wait a bit.
- Kill the app, preserving the state.
- Relaunch the app, with the state restoring to a Reader stream.
- Check that the cell layouts are exact, test by following the same steps above with screenshots, if needed.

:godmode: @frosty

kurzee commented:
Also, please test the new refactor in a81c3fa9e0669f91ae871ce094130ec7a13db0fc, this is a performance optimization that I should have done before but decided to put in now... I tested this myself after refactoring, but doesn't hurt to have more testing:

- Test by loading cards that should have a comment and like button, with correct numbers (tap to the detail).
- Test by loading cards that should not have a comment or like button, such as Jetpack posts, RSS feed posts, or other non-WPCom posts.

frosty commented:
So:

* The constraint issue fixed itself as soon as I opened the storyboard 👍 
* State restoration seems to be working fine for me.
* Like and Comment buttons look good!
* I'm still seeing issues when performing your screenshotting tests:

### Screenshot 1

![simulator screen shot 20 oct 2016 10 36 50](https://cloud.githubusercontent.com/assets/4780/19555924/a74ba208-96b6-11e6-8b8c-d5edbe5b56e7.png)

### Screenshot 2

![simulator screen shot 20 oct 2016 10 37 03](https://cloud.githubusercontent.com/assets/4780/19555918/a0210694-96b6-11e6-83a2-47da8146f5fc.png)

* I also saw this the first time I went into discover. It hasn't happened again since, though (and the cells fixed themselves after being recycled):

![simulator screen shot 20 oct 2016 10 42 21](https://cloud.githubusercontent.com/assets/4780/19555939/bd433774-96b6-11e6-8729-f81a6bc8839d.png)

![simulator screen shot 20 oct 2016 10 42 14](https://cloud.githubusercontent.com/assets/4780/19555940/bd7148da-96b6-11e6-9e5c-d168060cb5b0.png)

Sorry!

kurzee commented:
>The follow button IBAction is showing up as not connected properly – might just need disconnecting and reconnecting.

Finally figured out what's going on here. There is a bug in IB for Xcode 8 that adds `with` to the IBAction selectors.

For example, resetting the connection for a method of `@IBAction func tappedTagButton(sender: UIButton)` ends up creating a connection for the selector of `tappedTagButtonWithSender:`. This also removes the "!" warning seen in IB, _however_, running the application with the newly created connection crashes since there isn't actually a method matching `tappedTagButtonWithSender:`.

That said, we'll leave the connections and methods alone right now, even though IB thinks there is a broken connection(s).

kurzee commented on WordPress/Classes/ViewRelated/Reader/ReaderPostCardCell.swift:
```diff
@@ -357,53 +291,54 @@ import WordPressShared
             titleLabel.attributedText = NSAttributedString(string: title, attributes: attributes)
-            titleLabelBottomConstraint.constant = titleLabelBottomConstraintConstant
-
+            if titleLabel.hidden {
```
This was originally for performance, but based on a false assumption. I'm removing these checks.

kurzee commented:
>Occasionally the loading view and the no results view are a bit off center, slightly too far down the screen, but I think this is an existing issue.

- Fixed in 1acbee058a443b8b0bf15bc36fa0a74e17a62563

>Search: One the iPad, the dropdown for past searches sticks out beyond the width of the new card content. It might be nice if this could be made flush left and right.

- For this one, I've opted to instead pull the search bar and drop down to the full width of the screen. This matches the rest of the app's use of the search bar as readability doesn't really work on `UISearchBar`. I've also updated the styling to match `WPStyleGuide.configureSearchBar()`.

kurzee commented:
I think... I think I've covered everything here! More testing on the commits/comments from today would be highly appreciated.

For performance, I think I've tweaked it to about as good as it's going to get. Our heavy hitters are indeed the featured image views but primarily the many views that are a part of these cells. I think moving forward it's best to approach performance with as few views as possible and the different layout possibilities divided among multiple cells and identifiers. The less views, less constraint updates, calculations, etc. == max performance.

@frosty, @aerych, @bummytime can you retest on your older devices one last time for performance?? 🙏  I think you'll find things a tad bit smoother, but still just a bit laggy on dequeue. However, the previous implementation's delays on `viewWillAppear` and reloads no longer exist, so this is probably more usable to the DAU, overall.

frosty commented:
@kurzee Awesome! Performance looks good to me on the few devices I've got available at the moment. I'm afraid I still have a couple of issues for you, though:

* The search bar search icon appearance doesn't quite match the search bars elsewhere in the app. You can remove the extra configuration in `ReaderSearchViewController.swift` (lines 141 - 146), as it's already set up via UIAppearance.
* I've managed to reproduce the attribution text getting squashed vertically on Discover posts. I'm using the iPad Pro 9.7 simulator, in landscape. Resize via multitasking to a 50 / 50 split, and then back to full size. Every time for me, the attribution label gets squashed on the first post visible. I also saw it happen without performing these steps, but this way seems to at least reproduce it for me every time.
* I'm still getting issues with the initial image sizes. It's only on the first time visiting the stream (e.g. Discover) after launching the app. After that, the issue doesn't reappear until I relaunch. It also _seems_ like it only happens in landscape. Here's an example of how it looks:

![reader-readability-image-size-recycling-1](https://cloud.githubusercontent.com/assets/4780/19596381/454937ac-9786-11e6-82fd-310970e1b9bb.png)

![reader-readability-image-size-recycling-2](https://cloud.githubusercontent.com/assets/4780/19596384/479a3fa6-9786-11e6-8607-b442e526f01b.png)

* I've seen images changing after rotation (e.g. landscape -> portrait -> back to landscape). Here's one example from iPad, and one from iPhone:

![reader-readability-image-rotation](https://cloud.githubusercontent.com/assets/4780/19596424/6edd2d6c-9786-11e6-8804-55173dea02db.png)

![reader-readability-image-rotation-2](https://cloud.githubusercontent.com/assets/4780/19596462/a330c6aa-9786-11e6-9a23-c44abac889db.png)

* Finally, I saw this happen once and I haven't been able to reproduce it yet:

![reader-readability-squashed-content](https://cloud.githubusercontent.com/assets/4780/19596569/3da92b0a-9787-11e6-8d7f-335764068317.png)

Sorry! 🙈

aerych commented:
> For this one, I've opted to instead pull the search bar and drop down to the full width of the screen. This matches the rest of the app's use of the search bar as readability doesn't really work on UISearchBar. I've also updated the styling to match WPStyleGuide.configureSearchBar().

I'm digging the full screen search bar. Good call :) 
One thing, the search icon has the wrong color.  It should be the same shade of grey as the button to clear the text.
<img width="772" alt="screen shot 2016-10-21 at 10 55 12 am" src="https://cloud.githubusercontent.com/assets/1435271/19604919/0ba84d52-977d-11e6-8880-9a7dc1b2912c.png">

aerych commented:
> For performance, I think I've tweaked it to about as good as it's going to get. 

Scrolling on the iPad 3 seemed quite a bit better during this round of testing. Yay :)

aerych commented:
I've noticed one tiny issue when refreshing the post list.  When I return to a busy list after some time has passed and pull to refresh, if the newly inserted posts include a gap marker the position of the content I'm currently viewing is shifted down (about 60 pixels) rather than staying stationary on screen. 

I've reproduced this multiple times and its only happening when a gap marker cell is displayed.  I'm not seeing it happen when I pull to refresh and a gap marker is *not* displayed.  I'm also not seeing it happen if I pull to refresh, then scroll down a bit, even when a gap marker is displayed. 

I've reproduced the behavior on an iPad 3, and the 12 inch iPad Pro

kurzee commented:
>The search bar search icon appearance doesn't quite match the search bars elsewhere in the app. You can remove the extra configuration in ReaderSearchViewController.swift (lines 141 - 146), as it's already set up via UIAppearance.

>I'm digging the full screen search bar. Good call :) 
One thing, the search icon has the wrong color. It should be the same shade of grey as the button to clear the text.

- I've removed the custom searchbar styling in b5aa4bd7dec762316e0da2803a2548b581750e0e

>I've managed to reproduce the attribution text getting squashed vertically on Discover posts. I'm using the iPad Pro 9.7 simulator, in landscape. Resize via multitasking to a 50 / 50 split, and then back to full size. Every time for me, the attribution label gets squashed on the first post visible. I also saw it happen without performing these steps, but this way seems to at least reproduce it for me every time.

- This should now be finally fixed in ea2f3c8151c48692c92d95b713703e8d7d0ec759, basically, I removed a couple layout passes that had been added over time that were interfering with the timing of layout while dequeueing.

kurzee commented:
I've added a new tweak that will help just a bit with dequeueing new cells (performance). This moves the fetching of image request auth tokens to the `ReaderStreamViewController` instead of the cells.

@aerych can you take a look at the changes in https://github.com/wordpress-mobile/WordPress-iOS/pull/6077/commits/915cb3c40a13741eaae5b70f8f5d1e1683c4e33a and let me know if anything stands out?

aerych commented on WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:
```diff
@@ -210,6 +206,9 @@ import WordPressComAnalytics
         // pull to refresh animation.
         view.userInteractionEnabled = readerTopic != nil
 
+        NSNotificationCenter.defaultCenter().addObserver(self, selector: #selector(defaultAccountDidChange(_:)), name: WPAccountDefaultWordPressComAccountChangedNotification, object: nil)
```
I'm wondering if we can remove the observer completely.  It was useful when the post list was the root vc of the reader as it was always recycled, but with the recent menu refresh that's no longer the case.  All post streams are discarded during login/logout.   I suppose it might depend on how we're handing the split view integration.

kurzee replied:
Hmm, maybe it would be best to observe it anyway? Maybe not make an assumption that it will never change while the VC is allocated.

aerych replied:
Certainly no harm leaving it as is.  We can always revisit after the split view work is done.

frosty replied:
If I remember correctly, the splitview implementation may completely tear down and recreate the splitviews on account change anyway.

kurzee commented:
>I've noticed one tiny issue when refreshing the post list. When I return to a busy list after some time has passed and pull to refresh, if the newly inserted posts include a gap marker the position of the content I'm currently viewing is shifted down (about 60 pixels) rather than staying stationary on screen.

This one is quite baffling. I thought I had figured out what's going on... I think we'll have to chase this one down separately. The issue is something to do with the gap cell and the `tableHeaderView`.

Regardless, I added another piece of logic that takes care of any cells in the calculation that were deleted, in 6b179fad82a773c4c3049ecf2829d128cd1c0727.

kurzee commented:
>This one is quite baffling. I thought I had figured out what's going on... I think we'll have to chase this one down separately. The issue is something to do with the gap cell and the tableHeaderView.

Hmm, well I think I've actually fixed this one with my last commit or something else today. Can't reproduce it now. Will test it some more over the weekend, either way, I think we can tackle this as a separate issue/PR if we nail it down.

Sample testing steps:

- Follow a test site of your own
- Edit `ReaderPostServiceNumberToSync` in `ReaderPostService` to be 4, for testing.
- Load your "Followed Sites" in the Reader on iPhone or iPad
- Create 6 or so posts on your test site
- Pull down to refresh, some of the posts and a gap should load
- The offset should stay focused on the original content offset and not jump

frosty commented:
Wow, just gave this a quick test on my iPad 3 (which I finally got working again). I'm really impressed by the performance! Nice job, @kurzee!

frosty commented:
@kurzee I'm afraid I'm still getting some image and label issues. Following my previous reproduction steps for the squashed label (go to 50/50 multitasking split and then back to full screen) I get this happening to the attribution label - reproducible every time:

<img width="348" alt="screen shot 2016-10-24 at 11 16 03" src="https://cloud.githubusercontent.com/assets/4780/19642042/452f9b56-99db-11e6-845d-247fa215e53c.png">

And I'm also still seeing issues with images changing when cells are recycled – I'm not sure if you tackled this yet or not? See below:

![reader-readability-image-recycling-1](https://cloud.githubusercontent.com/assets/4780/19642274/77050aca-99dc-11e6-838a-fa2a2b9d1eca.png)

![reader-readability-image-recycling-2](https://cloud.githubusercontent.com/assets/4780/19642273/76d6239a-99dc-11e6-877b-0a0e08f03847.png)

kurzee commented:
@frosty I think I've figured it out! Had a constraint on the attribution view that was actually throwing the whole layout off. Revising the constraint also allows us to not have to trigger any layout passes on configuration.

Also fixed the imageView issue, _did_ need to trigger a layout pass on that one before checking the frame width.

Can you give it another go?

frosty commented:
We're so close! I can confirm that I can't recreate the attribution label issue any more. The featured image issue also seems to have gone when first visiting the stream.

I'm still seeing one smaller issue with images, which I created by opening to a portrait split, and rotating to landscape and back again – as you can see, the visible portion of the images change:

![simulator screen shot 25 oct 2016 15 09 42](https://cloud.githubusercontent.com/assets/4780/19693724/6c1fa144-9ad4-11e6-8a67-9da512d5755b.png)

![simulator screen shot 25 oct 2016 15 09 42](https://cloud.githubusercontent.com/assets/4780/19693736/6ec42ca8-9ad4-11e6-916a-c1dd5d8e5990.png)

kurzee commented:
>I'm still seeing one smaller issue with images, which I created by opening to a portrait split, and rotating to landscape and back again – as you can see, the visible portion of the images change:

@frosty hmm, I can't reproduce this actually. Do you have specific steps to reproduce?

frosty commented:
@kurzee I've just been testing again and I can't now either. It was also a very minor issue – I think we should merge and we can always address it as a separate issue if it crops up again.

What do you think? Are we good to :shipit:?

kurzee commented:
@frosty WE'RE GOOD TO GO

kurzee commented:
@frosty thanks a ton for your excellent review and testing!! Really revealed some underlying issues, improving things substantially for the long-run.
